lib/Runtime/Types/RecyclableObject.h

```diff
@@ -466,4 +463,55 @@ namespace Js {
         int GetHeapEnumValidationCookie() { return m_heapEnumValidationCookie; }
 #endif
     };
+
+    // Return whether the given RecyclableObject is of the template parameter's type.
+    // Generally, subclasses of RecyclableObject should only need to provide
+    // a specialization for VarIs(RecyclableObject*), and the other conversion
+    // functions should take care of themselves.
+    template <typename T> bool VarIs(RecyclableObject* obj);
+
+    // Return whether the given Var is of the template parameter's type.
+    template <typename T> bool VarIs(Var aValue)
+    {
+        AssertMsg(aValue != nullptr, "VarIs: aValue is null");
+
+#if INT32VAR
+        bool isRecyclableObject = (((uintptr_t)aValue) >> VarTag_Shift) == 0;
+#else
+        bool isRecyclableObject = (((uintptr_t)aValue) & AtomTag) == AtomTag_Object;
+#endif
+
+        return isRecyclableObject && VarIs<T>(UnsafeVarTo<RecyclableObject>(aValue));
+    }
+
+    template <> inline bool VarIs<RecyclableObject>(RecyclableObject* obj)
+    {
+        return true;
+    }
+
+    CompileAssertMsg(AtomTag_Object == 0, "Ensure GC objects do not need to be marked");
+
+    template <typename T> T* VarTo(RecyclableObject* obj)
```

Comment:
It'd be nice to have this as
```cpp
template <typename T, typename U>
T* VarTo(U* obj)
{
    static_assert(!is_same<T,U>::value, "Check should be unnecessary - did you prematurely cast?");
    static_assert(is_base_of<U,T>::value, "VarTo should only downcast!");
    return static_cast<T*>(obj);
}
```

This way, we'd check for two more properties that could indicate logical issues in call sites.

Reply:
Do you have a recommendation for how to handle `TemplatedForEachItemInRange`? It has an input parameter of a templated type, and calls `VarTo<JavascriptArray>` on that input variable. For that call site, `is_same<T,U>` is sometimes true and sometimes false depending on the template parameter. Should we have a separate casting operation that allows `is_same`? `VarToOrNotReallyIKnowWhatImDoing`?

---
In reply to: [202229876](https://github.com/Microsoft/ChakraCore/pull/5456#discussion_r202229876) [](ancestors = 202229876)

Reply:
Potentially, or the TemplatedForEachItemInRange could have different a different implementation if is_same<T,U>.

Reply:
I looked back over TemplatedForEachItemInRange, and it's now apparent why it's an issue - we're using a pointer of the old type after it got mutated by the foreach'd function. We should likely re-write that function with something like this:

```cpp
template<typename T, typename P>
static BOOL TryTemplatedGetItem(Var arr, P index, Var *element, ScriptContext *scriptContext, bool checkHasItem = true)
{
    return T::Is(arr) ? JavascriptArray::TemplatedGetItem(VarToUnsafe<T>(arr), index, element, scriptContext, checkHasItem) :
        JavascriptOperators::GetItem(arr, index, element, scriptContext);
}

template <bool hasSideEffect, typename T, typename Fn>
static void TemplatedForEachItemInRange(T * arr, uint32 startIndex, uint32 limitIndex, Var missingItem, ScriptContext * scriptContext, Fn fn)
{
    for (uint32 i = startIndex; i < limitIndex; i++)
    {
        Var element;
        Var curArr = arr;
        fn(i, TryTemplatedGetItem<T>(arr, i, &element, scriptContext) ? element : missingItem);

        if (hasSideEffect && MayChangeType<T>() && !T::Is(curArr))
        {
            // The function has changed, go to another ForEachItemInRange. It is possible that the array might have changed to
            // an ES5Array, in such cases we don't need to call the JavascriptArray specific implementation.
            if (JavascriptArray::Is(curArr))
            {
                VarTo<JavascriptArray>(curArr)->template ForEachItemInRange<true>(i + 1, limitIndex, missingItem, scriptContext, fn);
                return;
            }
            else
            {
                AssertOrFailFastMsg(VarIs<ES5Array>(curArr), "The array should have been converted to an ES5Array");
            }
        }
    }
}

template <bool hasSideEffect, typename T, typename P, typename Fn>
static void TemplatedForEachItemInRange(T * arr, P startIndex, P limitIndex, ScriptContext * scriptContext, Fn fn)
{
    for (P i = startIndex; i < limitIndex; i++)
    {
        Var element;
        Var curArr = arr;
        if (TryTemplatedGetItem<T>(curArr, i, &element, scriptContext))
        {
            fn(i, element);

            if (hasSideEffect && MayChangeType<T>() && !T::Is(curArr))
            {
                // The function has changed, go to another ForEachItemInRange. It is possible that the array might have changed to
                // an ES5Array, in such cases we don't need to call the JavascriptArray specific implementation.
                if (JavascriptArray::Is(curArr))
                {
                    VarTo<JavascriptArray>(curArr)->template ForEachItemInRange<true>(i + 1, limitIndex, scriptContext, fn);
                    return;
                }
                else
                {
                    AssertOrFailFastMsg(VarIs<ES5Array>(curArr), "The array should have been converted to an ES5Array");
                }
            }
        }
    }
}
```

This patterning makes me more convinced that we should have those static asserts, since this is a (thankfully safe in this case) type confusion opportunity.

Reply:
Thanks @PenguinWizzard for the suggested fix; it works great! I used `RecyclableObject*` rather than `Var` to ensure we weren't slowing anything down due to tag checks. Does that seem reasonable?

---
In reply to: [203209057](https://github.com/Microsoft/ChakraCore/pull/5456#discussion_r203209057) [](ancestors = 203209057)

Reply:
I'll open a new comment chain on the new change about that; there's one concern that I have with it.